### PR TITLE
Solution asset cache refactoring

### PR DESCRIPTION
**Pull request by tmat**
Refactor AssetStorage on the server side.

- Renames AssetStorage to SolutionAssetCache.
- Remove IAssetSource from SolutionAssetCache as it it snot used by the type itself.
- Move the default instance from ServiceBase to RemoteWorkspaceManager.

### CODE REVIEW

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal partial class RemoteWorkspace
+    {
+        /// <summary>
+        /// Create solution for given checksum from base solution
+        /// </summary>
+        private readonly struct SolutionCreator
```
Comment (tmat) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:28]: >private readonly struct SolutionCreator [](start = 8, length = 39)

Moved to a nested class.

File: src/Workspaces/Remote/ServiceHub/Host/SolutionCreator.cs [hunk 2]
```diff
@@ -1,534 +0,0 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/SolutionCreator.cs:1]: in general, i recommend not renamign the file in this main PR.  this makes the diff easier.  then just have a quick followup pr that renames.

File: src/Workspaces/Remote/ServiceHub/Host/SolutionCreator.cs [hunk 3]
```diff
@@ -1,534 +0,0 @@
...
-            async Task<Solution> CreateSolutionFromScratchAsync(Checksum checksum)
-            {
-                var (solutionInfo, options) = await _assetProvider.CreateSolutionInfoAndOptionsAsync(checksum, _cancellationToken).ConfigureAwait(false);
-                var workspace = new TemporaryWorkspace(_hostServices, WorkspaceKind.RemoteTemporaryWorkspace, solutionInfo, options);
-                return workspace.CurrentSolution;
-            }
-        }
-#endif
-    }
-}
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/SolutionCreator.cs:534]: trusting that nothing changed here :)

File: src/Workspaces/Remote/ServiceHub/Host/SolutionCreator.cs (hunk 2 above)
Comment (tmat) [reply src/Workspaces/Remote/ServiceHub/Host/SolutionCreator.cs:1]: Tried to make it easier in a separate commit.

File: src/Workspaces/Remote/ServiceHub/Host/SolutionCreator.cs (hunk 3 above)
Comment (tmat) [reply src/Workspaces/Remote/ServiceHub/Host/SolutionCreator.cs:534]: Yup, no sig changes.

**Pull request merged**